bcm2835-unicam driver for unicam

naushir opened the pull request:
This PR adds the existing (in the 4.19.y kernel tree) bcm2835-unicam as well as adding support for an embedded data node in the driver.

Modify the imx219 driver to expose embedded data on a new pad.

naushir committed:
media: uapi: Add MEDIA_BUS_FMT_SENSOR_DATA media bus format

This patch adds MEDIA_BUS_FMT_SENSOR_DATA used by the bcm2835-unicam
driver to support CSI-2 embedded data streams from camera sensors.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

6by9 commented:
This needs a rebase. With the upstream branch being rebased it's impossible to review like this.

6by9 commented on drivers/media/platform/bcm2835/vc4-regs-unicam.h:
```diff
@@ -0,0 +1,266 @@
+ * Copyright (C) 2017 Raspberry Pi Trading.
+ * Dave Stevenson <dave.stevenson@raspberrypi.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
```
We've got SPDX headers, so shouldn't have the boilerplate

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2100 @@
+ * not be able to stream data.
+ *
+ *
+ * This program is free software; you may redistribute it and/or modify
```
We've got the SPDX header, so shouldn't have the boilerplate

naushir replied:
Done.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1953,6 +1958,16 @@ static int unicam_probe_complete(struct unicam_device *unicam)
 		return ret;
 	}
 
+	ret = media_create_pad_link(&unicam->sensor->entity, 0,
```
As per the discussion on the libcamera list, the call above to v4l2_device_register_subdev_nodes needs to be removed to avoid apps being able to reconfigure the source device indirectly.

6by9 replied:
Except now Jacopo has pushed patches to add a read-only subdev node, so download the series from patchwork (https://patchwork.linuxtv.org/patch/62630/) and apply (add a comment over the source to the commit text) before our patches.

naushir replied:
Done

naushir commented:
@6by9, can you have a first pass at reviewing this please?  All staggered write functionality has been removed for now, and will be added as another PR.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -683,6 +683,18 @@ static void unicam_wr_dma_addr(struct unicam_device *dev, unsigned int dmaaddr,
 
+static inline int unicam_get_lines_done(struct unicam_device *dev)
+{
+	u32 start_addr, cur_addr;
```
dma_addr_t, not u32.

naushir replied:
Done

6by9 commented:
Please can you do a rebase on this? It's not possible to review the overall changes due to needing a rebase, and there are a few too many patches at present to review them individually.

Squash all Unicam driver patches from "media: bcm2835-unicam: Driver for CCP2/CSI2 camera interface" to "media: bcm2835-unicam: Remove any requirement to align the height" as we don't care about the history of those really. "media: bcm2835-unicam: Set stride alignment to 32 bytes. " can be squashed in as well.
The only bits that really want to stay as independent patches are those that add the extra pad/node and metadata support (the potentially contentious bits).

"media: bcm2835-unicam: Add embedded data node." is before "media: uapi: v4l2-core: Add embedded data V4L2 foucc type.", and "media: uapi: Add MEDIA_BUS_FMT_CSI2_EMBEDDED_DATA media bus format". Do the intermediate commits build then, because I'd expect them to be missing the defines. Please ensure they are sensibly ordered.

Any new format really requires documentation too. https://www.kernel.org/doc/html/latest/media/uapi/v4l/meta-formats.html which is rendered from Documentation/media/uapi/v4l/

6by9 reviewed (commented):
I started reviewing and had some comments pending. I think that was when I queried whether we had all the fixups that had been previously requested.

naushir commented:
> Please can you do a rebase on this? It's not possible to review the overall changes due to needing a rebase, and there are a few too many patches at present to review them individually.
> 
> Squash all Unicam driver patches from "media: bcm2835-unicam: Driver for CCP2/CSI2 camera interface" to "media: bcm2835-unicam: Remove any requirement to align the height" as we don't care about the history of those really. "media: bcm2835-unicam: Set stride alignment to 32 bytes. " can be squashed in as well.
> The only bits that really want to stay as independent patches are those that add the extra pad/node and metadata support (the potentially contentious bits).
> 

Will do.

> "media: bcm2835-unicam: Add embedded data node." is before "media: uapi: v4l2-core: Add embedded data V4L2 foucc type.", and "media: uapi: Add MEDIA_BUS_FMT_CSI2_EMBEDDED_DATA media bus format". Do the intermediate commits build then, because I'd expect them to be missing the defines. Please ensure they are sensibly ordered.
> 

Not entirely sure why, but the commit list ordering above differs from the actual branch https://github.com/naushir/linux/commits/5.4-unicam...?  On the branch, the commits are ordered as you would expect, i.e. allow successful builds at each commit.

naushir commented:
@6by9, changes rebased/squashed/updated.

> Any new format really requires documentation too. https://www.kernel.org/doc/html/latest/media/uapi/v4l/meta-formats.html which is rendered from Documentation/media/uapi/v4l/

This is still pending, but I thought I'd give you a chance to review the other changes first.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -56,6 +56,7 @@
 #include <linux/slab.h>
 #include <linux/uaccess.h>
 #include <linux/videodev2.h>
+#include <linux/dma-mapping.h>
```
Includes should be in alphabetical order.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -682,14 +694,17 @@ static void unicam_wr_dma_addr(struct unicam_device *dev, unsigned int dmaaddr,
+				      0);
 	cur_addr = reg_read(&dev->cfg, UNICAM_IBWP);
-	return (cur_addr - start_addr) / stride;
+	return (u32)(cur_addr - start_addr) / stride;
```
The prototype is to return int, but you've cast to a u32.
Defined size types (eg u32) should only be used for register access type operations. I'd expect this to return unsigned int.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -386,6 +393,11 @@ struct unicam_node {
 	struct v4l2_ctrl_handler ctrl_handler;
 	unsigned int embedded_lines;
+	/* Dummy buffer intended to dump unicam data
+	 * if we have no queued buffers to swap to.
```
"to dump unicam data *to* if" or "to take unicam data if". It seems incorrect as currently written.

naushir replied:
Done

6by9 commented on include/uapi/linux/media-bus-format.h:
```diff
@@ -144,6 +144,9 @@
 #define MEDIA_BUS_FMT_SGRBG16_1X16		0x301f
 #define MEDIA_BUS_FMT_SRGGB16_1X16		0x3020
 
+/* CSI2 Camera sensor embedded metadata - next is 0x4001 */
```
"next is 0x4001"??! What/where/why/how?
a) This shouldn't be in the block with Bayer as it isn't Bayer.
b) If it were Bayer then the comment at the top of the block needed to be updated as the next would be 0x3022, and the defines are in alphabetical order.

Add at the very bottom as
```
/* Metadata formats - next is 0x7002 */
#define MEDIA_BUS_FMT_CSI2_EMBEDDED_DATA	0x7001
```

naushir replied:
Done

naushir replied:
Actually, just going to backtrack on this one.  In the header, it states:

`/* Vendor specific formats - next is	0x5002 */`

CSI2 embedded data is a general format, so perhaps I should ~~set it to value 0x5001 and move the vendor specific formats further down by 0x1000?~~ retain the original change and be explicit that it is not a bayer format?

6by9 replied:
You can't change the existing defines as they are part of the uapi and are therefore published to the world.

0x5001-0x5fff are vendor formats, which is one odd YUYV format.
0x6001-0x6fff are HSV generic formats.
Adding Metadata formats starting at 0x7001 is therefore valid. 

The group should be called "Metadata formats". Your specific one is for CSI2 data. Is embedded metadata not valid on CCP2 and CSI1 as well?
And who has actually defined the format? Is it MIPI or SMIA? Do we want to differentiate there?

naushir replied:
> Is embedded metadata not valid on CCP2 and CSI1 as well?
> And who has actually defined the format? Is it MIPI or SMIA? Do we want to differentiate there?

I do not believe CCP2 and CSI1 support embedded data (I could not find anything from a quick google search).  From what I know, SMIA defined the ED structure for the CSI-2 spec, I've only ever know it to be called CSI-2 Embedded Data, but perhaps SMIA Embedded Data would be a suitable term? I don't know...

With other sensor vendors, e.g. Omnivision, the format is entirely different, and a new format  type will certainly have to be defined.

6by9 commented:
I'm guessing Github is putting the commits in chronological order in the PR - not so helpful!

Sorry, more reordering/squashing.
The use of v4l2_device_register_ro_subdev_nodes should be in the base driver commit. There's nothing that needs to be kept separate there.
Same with "Do not write new buffer address on Frame End". I don't see a reason that needs to be separate.

More review comments on the specific patches in a bit. The main patch needs a full read through.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -488,6 +496,8 @@ static int check_mbus_format(struct unicam_device *dev,
 		memset(&mbus_code, 0, sizeof(mbus_code));
 		mbus_code.index = i;
+		mbus_code.pad = IMAGE_PAD;
+		mbus_code.which = V4L2_SUBDEV_FORMAT_ACTIVE;
```
Why have we gained setting ```mbus_code.which = V4L2_SUBDEV_FORMAT_ACTIVE;``` in this patch? Surely it was needed before, therefore should have been in the base driver patch.

naushir replied:
Done.

6by9 commented on arch/arm/configs/bcm2711_defconfig:
```diff
@@ -850,6 +850,7 @@ CONFIG_VIDEO_EM28XX_V4L2=m
 CONFIG_VIDEO_EM28XX_ALSA=m
 CONFIG_VIDEO_EM28XX_DVB=m
 CONFIG_V4L_PLATFORM_DRIVERS=y
+CONFIG_VIDEO_BCM2835_UNICAM=m
```
Adding to the defconfig should be in a different patch, and should be enabled in bcmrpi, bcm2708 and bcm2711 defconfigs. Possibly arch/arm64/configs/ variants too.

naushir replied:
Done.

I have left out arch/arm64.  I want to test this on 64-bit build before adding by default.

6by9 commented on MAINTAINERS:
```diff
@@ -3206,7 +3206,8 @@ F:	Documentation/devicetree/bindings/media/rpivid_hevc.jaml
 BROADCOM BCM2835 CAMERA DRIVER
-M:	Dave Stevenson <dave.stevenson@raspberrypi.org>
+M:	Dave Stevenson <dave.stevenson@raspberrypi.com>
+M:	Naushir Patuck <naush@raspberrypi.com>
```
We do now have kernel-list@raspberrypi.com for the HEVC driver, mainly as we don't want to drop everything on John.
Do we want to use it here instead of adding you?

6by9 replied:
The list of files will also change very soon to be the JAML version of the DT binding.

naushir replied:
Yes, that seems reasonable.  I've removed both names and added kernel-list@.

6by9 reviewed (commented):
Comments so far. I'll look at the main patch in detail later on.

And my Signed-off-by, and possibly authorship in the main driver patch still lists .org. Can we amend to .com please?

naushir committed:
media: bcm2835-unicam: Driver for CCP2/CSI2 camera interface

Add driver for the Unicam camera receiver block on
BCM283x processors.

This commit is made up of a series of changes cherry-picked from the
rpi-4.19.y branch.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: uapi: v4l2-core: Add sensor ancillary data V4L2 foucc type.

Add V4L2_META_FMT_SENSOR_DATA format 4CC.

This new format will be used by the BCM2835 Unicam device to return
out camera sensor embedded data.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
defconfig: Add bcm2835-unicam to platform defconfigs.

This commit enables the bcm2835-unicam driver in the following
defconfigs:

bcm2709_defconfig
bcm2711_defconfig
bcmrpi_defconfig

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

6by9 commented:
Cherry-picking them to my branch I get checkpatch issues. Adding checkpatch to your githooks really does mean you address them early ;-)
```
"media: bcm2835-unicam: Driver for CCP2/CSI2 camera interface"
CHECK: Please use a blank line after function/struct/union/enum declarations
#942: FILE: drivers/media/platform/bcm2835/bcm2835-unicam.c:857:
+}
+static int unicam_try_fmt_vid_cap(struct file *file, void *priv,

"media: uapi: v4l2-core: Add embedded data V4L2 foucc type."
ERROR: trailing statements should be on next line
#21: FILE: drivers/media/v4l2-core/v4l2-ioctl.c:1335:
+	case V4L2_META_FMT_EMBED_DATA:	descr = "CSI-2 Embedded data"; break;

"media: bcm2835-unicam: Use dummy buffer if none have been queued"
ERROR: trailing whitespace
#43: FILE: drivers/media/platform/bcm2835/bcm2835-unicam.c:396:
+^I/* $
```
There are also complaints about long lines, but generally it's becaue log lines have used V4L2_FOURCC_CONV and checkpatch doesn't cope with the split strings.

The patch adding V4L2_FOURCC_CONV has been rejected upstream, so we shouldn't be using it. Please revert to %08x for now, and we ought to revert 164737aafdbe7b14afe596b652ec2880f373dd0f (Odd I thought I'd used Sakari's version of that patch. Never mind).

"media: bcm2835-unicam: Use dummy buffer if none have been queued" ought to be squashed into the base driver commit, although it's not a straight squash as part needs to be become part of either "Add support for mulitple device nodes." or "Add embedded data node."
I'm not sure if I'm that fussed though.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+#define MAX_ENUM_MBUS_CODE	128
+
+/*
+ * Stride is a 16 bit register, but also has to be a multiple of 16.
```
Comment doesn't match defined value

naushir commented:
> There are also complaints about long lines, but generally it's becaue log lines have used V4L2_FOURCC_CONV and checkpatch doesn't cope with the split strings.
> 
> The patch adding V4L2_FOURCC_CONV has been rejected upstream, so we shouldn't be using it. Please revert to %08x for now, and we ought to revert [164737a](https://github.com/raspberrypi/linux/commit/164737aafdbe7b14afe596b652ec2880f373dd0f) (Odd I thought I'd used Sakari's version of that patch. Never mind).
> 

If I revert [164737a], I will also have to change bcm2835-v4l2-codec.c which uses the macros.   Should I include that change in a separate PR?

6by9 commented:
I thought et8ek8 did embedded data. Wasn't that a CCP2 sensor? Perhaps my
memory is failing.

> With other sensor vendors, e.g. Omnivision, the format is entirely
different, and a new format type will certainly have to be defined.

Drop it to _SENSOR_EMBEDDED_DATA?
It is very true that we don't want to have to add an extra format for
vendor-based formatting as we've already left it very vague with regard the
Bayer bit depth packing side of it.

6by9 reviewed (commented):
Part way through. I can't comment on existing threads without closing my review first!

6by9 commented:
> 
> 
> > There are also complaints about long lines, but generally it's becaue log lines have used V4L2_FOURCC_CONV and checkpatch doesn't cope with the split strings.
> > The patch adding V4L2_FOURCC_CONV has been rejected upstream, so we shouldn't be using it. Please revert to %08x for now, and we ought to revert [164737a](https://github.com/raspberrypi/linux/commit/164737aafdbe7b14afe596b652ec2880f373dd0f) (Odd I thought I'd used Sakari's version of that patch. Never mind).
> 
> If I revert 164737a, I will also have to change bcm2835-v4l2-codec.c which uses the macros. Should I include that change in a separate PR?

Don't worry about the revert for now then, but we do need to do it. Upstream's preference is to add a new printk %foo parser thing instead of macros.

6by9 commented:
And it looks like dom has just rebased the branch again to 5.4.29 :-(

naushir commented:
> Don't worry about the revert for now then, but we do need to do it. Upstream's preference is to add a new printk %foo parser thing instead of macros.

Would it be better if I make the change in this PR and add the commit for the codec driver as well?  I know it's not best to mix different changes in the PR, but this is a tiny one (only one printk line).

6by9 commented:
>>    Don't worry about the revert for now then, but we do need to do it. Upstream's preference is to add a new printk %foo parser thing instead of macros.

> Would it be better if I make the change in this PR and add the commit for the codec driver as well? I know it's not best to mix different changes in the PR, but this is a tiny one (only one printk line).

Seriously don't worry about it. As long as we drop it from the driver then I'm happy and it can potentially be upstreamed as is.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+
+	r = pm_runtime_get_sync(&dev->pdev->dev);
+
+	return r;
```
```return pm_runtime_get_sync(&dev->pdev->dev);```

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+/* Format setup functions */
+static const struct unicam_fmt *find_format_by_code(u32 code)
+{
+	unsigned int k;
```
i is the more common loop counter.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+static const struct unicam_fmt *find_format_by_pix(struct unicam_device *dev,
+						   u32 pixelformat)
+{
+	unsigned int k;
```
Again, i as loop counter?

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	int ret;
+
+	sd_fmt.which = V4L2_SUBDEV_FORMAT_ACTIVE;
+	sd_fmt.pad = 0;
```
sd_fmt has already been reset to 0.

6by9 replied:
And this could be squashed down to 
```
	struct v4l2_subdev_format sd_fmt = {
		.which = V4L2_SUBDEV_FORMAT_ACTIVE,
	};
```
same as __subdev_set_format below.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	if (ret < 0)
+		return ret;
+
+	*fmt = *mbus_fmt;
```
As this is the only use of mbus_fmt, it could be removed and done as a direct assignment from sd_fmt.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	struct v4l2_mbus_framefmt *mbus_fmt = &sd_fmt.format;
+	int ret;
+
+	*mbus_fmt = *fmt;
```
Again can we drop mbus_fmt?

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+{
+	unsigned int min_bytesperline;
+
+	v4l_bound_align_image(&f->fmt.pix.width, MIN_WIDTH, MAX_WIDTH, 2,
```
So we're aligning the width to a multiple of 2. Is that a reasonable requirement?

We do rely on that for computing YUYV format bytesperline, and a cropped Bayer format would be odd but surely plausible.
Conditional on being YUYV?

naushir replied:
I think this is a reasonable constraint.
Bayer ~will~ should never be odd (not sure our ISP will support odd widths on bayer input actually), and it would be extremely unlikely that any other formats will have odd widths.

6by9 replied:
RGB can be received via CSI, and the restriction on that to be even wouldn't really apply.
Leave it for now, and we can fix it up should it ever become a problem.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	}
+
+	v4l2_fill_pix_format(&dev->v_fmt.fmt.pix, &mbus_fmt);
+	dev->v_fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
```
This line feels like it's come from the splitting out for multiple streams.
It feels like it should be a value that is set at the start of day and never changes. I certainly see nothing in this function that would have changed it.

naushir replied:
Not sure I understand.  Did you mean that `dev->v_fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;` should be set once and only ever once (in probe() maybe) and never touched?

Logically this seems like the right place to put it - being the only(?) function that sets dev->v_fmt even though it may be a bit redundant.

naushir replied:
Well actually `dev->v_fmt` does get modified in `s_fmt`, which in-turn calls `reset_format()`

6by9 replied:
I was thinking probe or similar.  It is the slight funny of everything else being in v_fmt.pix (part of a union), and this is the lone member outside. If there isn't another suitable place, then leave it as is.

naushir replied:
I think it is probably the best logical place for this line.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	return 0;
+}
+
+static void unicam_wr_dma_addr(struct unicam_device *dev, unsigned int dmaaddr)
```
```dma_addr_t dmaaddr``` and do any casts explicitly in this function.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+static inline unsigned int unicam_get_lines_done(struct unicam_device *dev)
+{
+	dma_addr_t start_addr, cur_addr;
+	u32 stride = dev->v_fmt.fmt.pix.bytesperline;
```
unsigned int?
Whilst it's a u32 in the uapi, the maths we're doing on it here don't need to be fixed width.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	if (!dev->cur_frm)
+		return 0;
+
+	start_addr =
```
Indentation is horrid. I know the function name is huge which doesn't help.
```
struct unicam_buffer *frm = dev->cur_frm;

if (!frm)
  return 0;

start_addr = vb2_dma_contig_plane_dma_addr(&frm->vb.vb2_buf....
```
? It's shorter still than the multiple node version.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+			   dev->sensor->name, mbus_code.code, j);
+
+		fmt = find_format_by_code(mbus_code.code);
+		unicam_dbg(2, dev, "fmt %04x returned as %p, V4L2 FOURCC %04x, csi_dt %02X\n",
```
%04x? I thought %08x was more likely for an 4 ASCII char 4cc.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	struct unicam_device *dev = video_drvdata(file);
+	struct vb2_queue *q = &dev->buffer_queue;
+	const struct unicam_fmt *fmt;
+	struct v4l2_mbus_framefmt mbus_fmt = {0};
```
Upstream has a thing for longest variable def first, so swap fmt and mbus_fmt.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	struct v4l2_subdev_format sd_fmt = {
+		.which = V4L2_SUBDEV_FORMAT_TRY,
+	};
+	struct v4l2_mbus_framefmt *mbus_fmt = &sd_fmt.format;
```
Upstream has a thing for longest variable def first, so swap fmt to after mbus_fmt.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	dev->v_fmt.fmt.pix.bytesperline = f->fmt.pix.bytesperline;
+	unicam_reset_format(dev);
+
+	unicam_dbg(3, dev, "%s %dx%d, mbus_fmt %08X, V4L2 pix " V4L2_FOURCC_CONV ".\n",
```
As discussed, drop V4L2_FOURCC_CONV

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	struct unicam_dmaqueue *dma_queue = &dev->dma_queue;
+	unsigned long flags = 0;
+
+	/* recheck locking */
```
Odd comment (by me by the looks of it).
As long as all accesses to dma_queue are protected by the spinlock then this needs to be removed.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	spin_unlock_irqrestore(&dev->dma_queue_lock, flags);
+}
+
+static void unicam_wr_dma_config(struct unicam_device *dev,
```
It's called exactly once. Explicitly inline it? The compiler will anyway.
Either that or pull the write to IBLS back into unicam_start_rx. It's not worth the function call.

naushir replied:
Done (removed the function call)

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	}
+
+	val = 0;
+	set_field(&val, 2, UNICAM_DEBL_MASK);
```
We're not doing DPCM, so this can be left at 0.

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	    v4l2_subdev_has_op(dev->sensor, video, g_mbus_config)) {
+		struct v4l2_mbus_config mbus_config;
+
+		ret = v4l2_subdev_call(dev->sensor, video, g_mbus_config,
```
This needs separating out into a new commit and using g_mbus_format pad op that Jacopo introduces in https://patchwork.linuxtv.org/project/linux-media/list/?series=1945.
Needs a matching commit for tc358743, and possibly a munge of the adv728x overlay (I hate this bit!)

6by9 replied:
This comment still stands. I guess we merge it as is for now, but can't be upstreamed as is.

naushir replied:
Sorry, I actually missed this one.  Will leave it for now and merge as a separate change.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	video_set_drvdata(vdev, unicam);
+	vdev->entity.flags |= MEDIA_ENT_FL_DEFAULT;
+
+	ret = video_register_device(vdev, VFL_TYPE_GRABBER, -1);
```
Same bug as we had on bcm2835-codec.
```v4l2_disable_ioctl``` MUST be called before ```video_register_device``` - see https://elixir.bootlin.com/linux/latest/source/include/media/v4l2-dev.h#L448

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+	if (IS_ERR(unicam->clock)) {
+		unicam_err(unicam, "Failed to get clock\n");
+		return PTR_ERR(unicam->clock);
+	}
```
You'd talked of wanting a second clock so you could boost the core clock.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+
+module_platform_driver(unicam_driver);
+
+MODULE_AUTHOR("Dave Stevenson <dave.stevenson@raspberrypi.org>");
```
.com

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/vc4-regs-unicam.h:
```diff
@@ -0,0 +1,253 @@
+
+/*
+ * Copyright (C) 2017-2020 Raspberry Pi Trading.
+ * Dave Stevenson <dave.stevenson@raspberrypi.org>
```
.com

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+ *
+ * Copyright (C) 2017-2020 - Raspberry Pi (Trading) Ltd.
+ *
+ * Dave Stevenson <dave.stevenson@raspberrypi.org>
```
.com

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * BCM2835 Unicam capture Driver
```
Capitalisation

naushir replied:
Done

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+ *
+ * Dave Stevenson <dave.stevenson@raspberrypi.org>
+ *
+ * Based on TI am437x driver by Benoit Parrot and Lad, Prabhakar and
```
Indian name - Prabhakar Lad.
Amend to 
```
* Based on TI am437x driver by 
*   Benoit Parrot <bparrot@ti.com>
*   Lad, Prabhakar <prabhakar.csengg@gmail.com>
 ```

Likewise the line below ```Benoit Parrot, <bparrot@ti.com>```

naushir replied:
Done

6by9 reviewed (commented):
Sorry, I know a load of these are mine.

Build warnings too on the overall patchset:
```
In file included from /home/dave/Pi/linux2711/drivers/media/platform/bcm2835/bcm2835-unicam.c:61:0:
/home/dave/Pi/linux2711/drivers/media/platform/bcm2835/bcm2835-unicam.c: In function 'unicam_isr':
/home/dave/Pi/linux2711/include/media/v4l2-common.h:57:2: warning: 'ista' may be used uninitialized in this function [-Wmaybe-uninitialized]
  printk(level "%s: " fmt, (dev)->name , ## arg)
  ^
/home/dave/Pi/linux2711/drivers/media/platform/bcm2835/bcm2835-unicam.c:793:6: note: 'ista' was declared here
  int ista, sta;
      ^
In file included from /home/dave/Pi/linux2711/drivers/media/platform/bcm2835/bcm2835-unicam.c:61:0:
/home/dave/Pi/linux2711/include/media/v4l2-common.h:57:2: warning: 'sta' may be used uninitialized in this function [-Wmaybe-uninitialized]
  printk(level "%s: " fmt, (dev)->name , ## arg)
  ^
/home/dave/Pi/linux2711/drivers/media/platform/bcm2835/bcm2835-unicam.c:793:12: note: 'sta' was declared here
  int ista, sta;
            ^
```

6by9 commented:
The multiple node stuff also needs some revision. I can't stream /dev/video0 via v4l2-ctl, as it claims not all nodes are streaming. It also then throws up a couple of WARNs on ctrl-c.
```
[ 1104.105714] unicam fe801000.csi: Not all nodes are streaming yet.
[ 1110.323662] ------------[ cut here ]------------
[ 1110.323695] WARNING: CPU: 1 PID: 983 at /home/dave/Pi/linux2711/drivers/clk/clk.c:954 clk_core_disable+0x21c/0x288
[ 1110.323703] cam1 already disabled
[ 1110.323710] Modules linked in: rfcomm bnep hci_uart btbcm serdev bluetooth ecdh_generic ecc libaes 8021q garp stp llc binfmt_misc evdev imx219 bcm2835_v4l2(C) videobuf2_vmalloc bcm2835_codec(C) v4l2_mem2mem bcm2835_mmal_vchiq(C) vc_sm_cma(C) rpivid_mem brcmfmac brcmutil sha256_generic libsha256 v3d gpu_sched cfg80211 bcm2835_unicam rfkill videobuf2_dma_contig v4l2_dv_timings v4l2_fwnode videobuf2_memops videobuf2_v4l2 raspberrypi_hwmon hwmon videobuf2_common vc4 videodev cec mc drm_kms_helper i2c_bcm2835 drm drm_panel_orientation_quirks snd_soc_core snd_compress snd_pcm_dmaengine snd_pcm snd_timer snd syscopyarea sysfillrect sysimgblt fb_sys_fops uio_pdrv_genirq uio sg i2c_dev ip_tables x_tables ipv6 nf_defrag_ipv6
[ 1110.323885] CPU: 1 PID: 983 Comm: v4l2-ctl Tainted: G         C        5.4.29-v7l+ #15
[ 1110.323893] Hardware name: BCM2711
[ 1110.323900] Backtrace: 
[ 1110.323921] [<c020d51c>] (dump_backtrace) from [<c020d81c>] (show_stack+0x20/0x24)
[ 1110.323934]  r6:efa28000 r5:00000000 r4:c129abf8 r3:1d40f1a9
[ 1110.323951] [<c020d7fc>] (show_stack) from [<c0a26dd0>] (dump_stack+0xe0/0x124)
[ 1110.323968] [<c0a26cf0>] (dump_stack) from [<c0221da4>] (__warn+0xec/0x104)
[ 1110.323980]  r8:000003ba r7:00000009 r6:c0df8708 r5:00000000 r4:efa29cbc r3:1d40f1a9
[ 1110.323995] [<c0221cb8>] (__warn) from [<c0221e44>] (warn_slowpath_fmt+0x88/0xc4)
[ 1110.324007]  r9:c0df8708 r8:000003ba r7:c06e641c r6:00000009 r5:c0df8a48 r4:efa28000
[ 1110.324023] [<c0221dc0>] (warn_slowpath_fmt) from [<c06e641c>] (clk_core_disable+0x21c/0x288)
[ 1110.324035]  r9:ef2c3408 r8:ef2c3548 r7:ef2c32f0 r6:ef2c3560 r5:ef2aca80 r4:ef2aca80
[ 1110.324051] [<c06e6200>] (clk_core_disable) from [<c06e64b0>] (clk_core_disable_lock+0x28/0x34)
[ 1110.324061]  r6:ef2c3560 r5:ef2aca80 r4:80000013
[ 1110.324077] [<c06e6488>] (clk_core_disable_lock) from [<c06e64e8>] (clk_disable+0x2c/0x30)
[ 1110.324085]  r5:cec6ae40 r4:ef2c3040
[ 1110.324119] [<c06e64bc>] (clk_disable) from [<bf3367cc>] (unicam_stop_streaming+0x9c/0x1a8 [bcm2835_unicam])
[ 1110.324164] [<bf336730>] (unicam_stop_streaming [bcm2835_unicam]) from [<bf1eefcc>] (__vb2_queue_cancel+0x38/0x288 [videobuf2_common])
[ 1110.324176]  r10:00000008 r9:ef2c3408 r8:ef2c3548 r7:cf1c2540 r6:ef2c3560 r5:ef2c3468
[ 1110.324185]  r4:ef2c3408 r3:bf336730
[ 1110.324234] [<bf1eef94>] (__vb2_queue_cancel [videobuf2_common]) from [<bf1f1fa0>] (vb2_core_queue_release+0x28/0x48 [videobuf2_common])
[ 1110.324247]  r10:00000008 r9:cf236e78 r8:ef2c3548 r7:cf1c2540 r6:ef2c3560 r5:ef2c3468
[ 1110.324255]  r4:ef2c3408 r3:ef805880
[ 1110.324298] [<bf1f1f78>] (vb2_core_queue_release [videobuf2_common]) from [<bf202adc>] (_vb2_fop_release+0x74/0x94 [videobuf2_v4l2])
[ 1110.324307]  r5:cf1c2540 r4:00000000
[ 1110.324341] [<bf202a68>] (_vb2_fop_release [videobuf2_v4l2]) from [<bf336424>] (unicam_release+0x4c/0xd0 [bcm2835_unicam])
[ 1110.324352]  r6:cf264c40 r5:00000000 r4:ef2c32f0 r3:cf335c80
[ 1110.324452] [<bf3363d8>] (unicam_release [bcm2835_unicam]) from [<bf2605fc>] (v4l2_release+0xbc/0xe8 [videodev])
[ 1110.324464]  r8:ce81c770 r7:cf400790 r6:cf236e78 r5:cf1c2540 r4:ef2c3560 r3:bf3363d8
[ 1110.324558] [<bf260540>] (v4l2_release [videodev]) from [<c03fb4a0>] (__fput+0xa4/0x238)
[ 1110.324567]  r5:000e001f r4:cf1c2540
[ 1110.324584] [<c03fb3fc>] (__fput) from [<c03fb64c>] (____fput+0x18/0x1c)
[ 1110.324595]  r10:efa29f58 r9:cf1c2540 r8:c12b4530 r7:ce7f7008 r6:ce7f6ac0 r5:ce7f6ff0
[ 1110.324603]  r4:cf5839c0
[ 1110.324621] [<c03fb634>] (____fput) from [<c0242ea0>] (task_work_run+0xb8/0xdc)
[ 1110.324639] [<c0242de8>] (task_work_run) from [<c0226470>] (do_exit+0x3c4/0xb2c)
[ 1110.324651]  r9:cf459e40 r8:efa29f44 r7:00000001 r6:00000000 r5:cf459e00 r4:ce7f6ac0
[ 1110.324667] [<c02260ac>] (do_exit) from [<c0226c70>] (do_group_exit+0x4c/0xc4)
[ 1110.324675]  r7:c1204d88
[ 1110.324692] [<c0226c24>] (do_group_exit) from [<c0234290>] (get_signal+0x1a0/0xa74)
[ 1110.324702]  r6:efa28000 r5:00000009 r4:00000008 r3:efa29f44
[ 1110.324717] [<c02340f0>] (get_signal) from [<c020cbdc>] (do_work_pending+0x290/0x500)
[ 1110.324729]  r10:b6b6851c r9:b6b68518 r8:efa29f44 r7:c1204d88 r6:efa28000 r5:efa29fb0
[ 1110.324736]  r4:efa28000
[ 1110.324749] [<c020c94c>] (do_work_pending) from [<c0201034>] (slow_work_pending+0xc/0x20)
[ 1110.324758] Exception stack(0xefa29fb0 to 0xefa29ff8)
[ 1110.324769] 9fa0:                                     00000003 c0445611 beaa5018 beaa4dc0
[ 1110.324781] 9fc0: 00073b88 00000001 00000001 00000036 c0445611 00075520 c044560f 0004a164
[ 1110.324792] 9fe0: 00073ec0 beaa4da4 00014110 b6b68518 20000010 00000003
[ 1110.324805]  r10:00000036 r9:efa28000 r8:c02011c4 r7:00000036 r6:00000001 r5:00000001
[ 1110.324812]  r4:00073b88
[ 1110.324821] ---[ end trace a83b5097dababa16 ]---
[ 1110.324832] ------------[ cut here ]------------
[ 1110.324853] WARNING: CPU: 1 PID: 983 at /home/dave/Pi/linux2711/drivers/clk/clk.c:813 clk_core_unprepare+0x20c/0x238
[ 1110.324865] cam1 already unprepared
[ 1110.324876] Modules linked in: rfcomm bnep hci_uart btbcm serdev bluetooth ecdh_generic ecc libaes 8021q garp stp llc binfmt_misc evdev imx219 bcm2835_v4l2(C) videobuf2_vmalloc bcm2835_codec(C) v4l2_mem2mem bcm2835_mmal_vchiq(C) vc_sm_cma(C) rpivid_mem brcmfmac brcmutil sha256_generic libsha256 v3d gpu_sched cfg80211 bcm2835_unicam rfkill videobuf2_dma_contig v4l2_dv_timings v4l2_fwnode videobuf2_memops videobuf2_v4l2 raspberrypi_hwmon hwmon videobuf2_common vc4 videodev cec mc drm_kms_helper i2c_bcm2835 drm drm_panel_orientation_quirks snd_soc_core snd_compress snd_pcm_dmaengine snd_pcm snd_timer snd syscopyarea sysfillrect sysimgblt fb_sys_fops uio_pdrv_genirq uio sg i2c_dev ip_tables x_tables ipv6 nf_defrag_ipv6
[ 1110.325117] CPU: 1 PID: 983 Comm: v4l2-ctl Tainted: G        WC        5.4.29-v7l+ #15
[ 1110.325125] Hardware name: BCM2711
[ 1110.325132] Backtrace: 
[ 1110.325148] [<c020d51c>] (dump_backtrace) from [<c020d81c>] (show_stack+0x20/0x24)
[ 1110.325159]  r6:efa28000 r5:00000000 r4:c129abf8 r3:1d40f1a9
[ 1110.325174] [<c020d7fc>] (show_stack) from [<c0a26dd0>] (dump_stack+0xe0/0x124)
[ 1110.325188] [<c0a26cf0>] (dump_stack) from [<c0221da4>] (__warn+0xec/0x104)
[ 1110.325200]  r8:0000032d r7:00000009 r6:c0df8708 r5:00000000 r4:efa29cd4 r3:1d40f1a9
[ 1110.325214] [<c0221cb8>] (__warn) from [<c0221e44>] (warn_slowpath_fmt+0x88/0xc4)
[ 1110.325226]  r9:c0df8708 r8:0000032d r7:c06e72dc r6:00000009 r5:c0df8ac4 r4:efa28000
[ 1110.325241] [<c0221dc0>] (warn_slowpath_fmt) from [<c06e72dc>] (clk_core_unprepare+0x20c/0x238)
[ 1110.325253]  r9:ef2c3408 r8:ef2c3548 r7:ef2c32f0 r6:ef2c3560 r5:cec6ae40 r4:ef2aca80
[ 1110.325268] [<c06e70d0>] (clk_core_unprepare) from [<c06e7530>] (clk_unprepare+0x34/0x3c)
[ 1110.325277]  r5:cec6ae40 r4:ef2aca80
[ 1110.325301] [<c06e74fc>] (clk_unprepare) from [<bf3367d4>] (unicam_stop_streaming+0xa4/0x1a8 [bcm2835_unicam])
[ 1110.325310]  r4:ef2c3040 r3:1d40f1a9
[ 1110.325351] [<bf336730>] (unicam_stop_streaming [bcm2835_unicam]) from [<bf1eefcc>] (__vb2_queue_cancel+0x38/0x288 [videobuf2_common])
[ 1110.325363]  r10:00000008 r9:ef2c3408 r8:ef2c3548 r7:cf1c2540 r6:ef2c3560 r5:ef2c3468
[ 1110.325371]  r4:ef2c3408 r3:bf336730
[ 1110.325417] [<bf1eef94>] (__vb2_queue_cancel [videobuf2_common]) from [<bf1f1fa0>] (vb2_core_queue_release+0x28/0x48 [videobuf2_common])
[ 1110.325429]  r10:00000008 r9:cf236e78 r8:ef2c3548 r7:cf1c2540 r6:ef2c3560 r5:ef2c3468
[ 1110.325437]  r4:ef2c3408 r3:ef805880
[ 1110.325478] [<bf1f1f78>] (vb2_core_queue_release [videobuf2_common]) from [<bf202adc>] (_vb2_fop_release+0x74/0x94 [videobuf2_v4l2])
[ 1110.325488]  r5:cf1c2540 r4:00000000
[ 1110.325520] [<bf202a68>] (_vb2_fop_release [videobuf2_v4l2]) from [<bf336424>] (unicam_release+0x4c/0xd0 [bcm2835_unicam])
[ 1110.325530]  r6:cf264c40 r5:00000000 r4:ef2c32f0 r3:cf335c80
[ 1110.325627] [<bf3363d8>] (unicam_release [bcm2835_unicam]) from [<bf2605fc>] (v4l2_release+0xbc/0xe8 [videodev])
[ 1110.325639]  r8:ce81c770 r7:cf400790 r6:cf236e78 r5:cf1c2540 r4:ef2c3560 r3:bf3363d8
[ 1110.325729] [<bf260540>] (v4l2_release [videodev]) from [<c03fb4a0>] (__fput+0xa4/0x238)
[ 1110.325739]  r5:000e001f r4:cf1c2540
[ 1110.325755] [<c03fb3fc>] (__fput) from [<c03fb64c>] (____fput+0x18/0x1c)
[ 1110.325767]  r10:efa29f58 r9:cf1c2540 r8:c12b4530 r7:ce7f7008 r6:ce7f6ac0 r5:ce7f6ff0
[ 1110.325774]  r4:cf5839c0
[ 1110.325791] [<c03fb634>] (____fput) from [<c0242ea0>] (task_work_run+0xb8/0xdc)
[ 1110.325808] [<c0242de8>] (task_work_run) from [<c0226470>] (do_exit+0x3c4/0xb2c)
[ 1110.325820]  r9:cf459e40 r8:efa29f44 r7:00000001 r6:00000000 r5:cf459e00 r4:ce7f6ac0
[ 1110.325835] [<c02260ac>] (do_exit) from [<c0226c70>] (do_group_exit+0x4c/0xc4)
[ 1110.325843]  r7:c1204d88
[ 1110.325859] [<c0226c24>] (do_group_exit) from [<c0234290>] (get_signal+0x1a0/0xa74)
[ 1110.325870]  r6:efa28000 r5:00000009 r4:00000008 r3:efa29f44
[ 1110.325884] [<c02340f0>] (get_signal) from [<c020cbdc>] (do_work_pending+0x290/0x500)
[ 1110.325896]  r10:b6b6851c r9:b6b68518 r8:efa29f44 r7:c1204d88 r6:efa28000 r5:efa29fb0
[ 1110.325903]  r4:efa28000
[ 1110.325916] [<c020c94c>] (do_work_pending) from [<c0201034>] (slow_work_pending+0xc/0x20)
[ 1110.325924] Exception stack(0xefa29fb0 to 0xefa29ff8)
[ 1110.325935] 9fa0:                                     00000003 c0445611 beaa5018 beaa4dc0
[ 1110.325947] 9fc0: 00073b88 00000001 00000001 00000036 c0445611 00075520 c044560f 0004a164
[ 1110.325958] 9fe0: 00073ec0 beaa4da4 00014110 b6b68518 20000010 00000003
[ 1110.325970]  r10:00000036 r9:efa28000 r8:c02011c4 r7:00000036 r6:00000001 r5:00000001
[ 1110.325978]  r4:00073b88
[ 1110.326060] ---[ end trace a83b5097dababa17 ]---
```

6by9 commented:
Having commented out the "all ports streaming" check, 2 compliance warnings
```
Required ioctls:
	test MC information (see 'Media Driver Info' above): OK
		warn: v4l2-compliance.cpp(634): media bus_info 'platform:fe801000.csi' differs from V4L2 bus_info 'platform:unicam fe801000.csi'

Allow for multiple opens:
		warn: v4l2-compliance.cpp(634): media bus_info 'platform:fe801000.csi' differs from V4L2 bus_info 'platform:unicam fe801000.csi'
```

One line fix to solve that....

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2372 @@
+		sizeof(unicam->mdev.model));
+	strscpy(unicam->mdev.serial, "", sizeof(unicam->mdev.serial));
+	snprintf(unicam->mdev.bus_info, sizeof(unicam->mdev.bus_info),
+		 "platform:%s", pdev->name);
```
```"platform:%s %s", pdev->dev.driver->name, dev_name(&pdev->dev));```
to avoid the v4l2-compliance warning. It then matches the combination of the snprintf in v4l2_device_register, plus the "platform" added by unicam_querycap

naushir replied:
Done

naushir replied:
Done

naushir commented:
> The multiple node stuff also needs some revision. I can't stream /dev/video0 via v4l2-ctl, as it claims not all nodes are streaming. It also then throws up a couple of WARNs on ctrl-c.

What command are you using for this?

6by9 commented:
> What command are you using for this?

```v4l2-ctl --stream-mmap=3 --stream-count=100 --stream-to=/dev/null```
It just stalls waiting for the other node to start streaming.

naushir commented:
Hmmm... I'm sure I had fixed that many moons back.  Oh well, probably lost it with the churn.  Never mind, should be an easy fix.

6by9 commented:
I'll agree that the code in unicam_all_nodes_streaming would appear to cover that case, but it didn't work. I don't know if v4l2-ctl opens other device nodes (it shouldn't really).

I haven't run v4l2-compliance -f which tests all the formats stream properly. That's more a test of the sensor driver as much as the receiver, although it would exercise the unpacking too.

naushir commented:
> I'll agree that the code in unicam_all_nodes_streaming would appear to cover that case, but it didn't work. I don't know if v4l2-ctl opens other device nodes (it shouldn't really).

Curious, it seems that something (during boot) is opening both nodes and then never closing them.  So the driver is waiting on the other node (embedded data) to start streaming.

This certainly used to work.  Not sure what could have changed other than some v4l2 core code.

6by9 commented:
I'll agree. Adding a ```dumpstack();``` to unicam_open and unicam_release it appears that something is opening and never closing (seemingly from userspace), but lsof isn't listing an open file handle.
```
[    8.386172] [<c020d51c>] (dump_backtrace) from [<c020d81c>] (show_stack+0x20/0x24)
[    8.386186]  r6:cedaa000 r5:00000000 r4:c129abf8 r3:9591b51d
[    8.386204] [<c020d7fc>] (show_stack) from [<c0a26dd0>] (dump_stack+0xe0/0x124)
[    8.386253] [<c0a26cf0>] (dump_stack) from [<bf2b04e0>] (unicam_open+0x34/0x12c [bcm2835_unicam])
[    8.386267]  r8:cf433040 r7:cf433b98 r6:cf7a6480 r5:cf7a6480 r4:cf433940 r3:bf1e7740
[    8.386379] [<bf2b04ac>] (unicam_open [bcm2835_unicam]) from [<bf1bf6d4>] (v4l2_open+0xac/0x118 [videodev])
[    8.386393]  r8:cf23aae8 r7:cf32d300 r6:cf433ca0 r5:cf7a6480 r4:cf433bb0 r3:bf2b04ac
[    8.386486] [<bf1bf628>] (v4l2_open [videodev]) from [<c03ff3d0>] (chrdev_open+0xf4/0x1c4)
[    8.386498]  r6:c1332a2c r5:cedaa000 r4:bf1da9dc r3:bf1bf628
[    8.386513] [<c03ff2dc>] (chrdev_open) from [<c03f57dc>] (do_dentry_open+0x244/0x3fc)
[    8.386528]  r10:cf7a6480 r9:c03ff2dc r8:cf7a6488 r7:00000000 r6:00000000 r5:cf23aae8
[    8.386536]  r4:cf7a6480
[    8.386552] [<c03f5598>] (do_dentry_open) from [<c03f6bf4>] (vfs_open+0x3c/0x40)
[    8.386565]  r9:cedabe90 r8:cedabf50 r7:00000000 r6:00000000 r5:00000000 r4:00020000
[    8.386582] [<c03f6bb8>] (vfs_open) from [<c040a560>] (path_openat+0x3ec/0x13c8)
[    8.386596] [<c040a174>] (path_openat) from [<c040c688>] (do_filp_open+0x7c/0xe8)
[    8.386609]  r10:00000142 r9:cedaa000 r8:00000001 r7:cedabe90 r6:cedabf50 r5:cedaa000
[    8.386617]  r4:cedaa000
[    8.386638] [<c040c60c>] (do_filp_open) from [<c03f6ed0>] (do_sys_open+0x138/0x204)
[    8.386650]  r8:c02011c4 r7:cf4ea000 r6:ffffff9c r5:cedaa000 r4:00000003
[    8.386664] [<c03f6d98>] (do_sys_open) from [<c03f6fe8>] (sys_openat+0x1c/0x20)
[    8.386677]  r10:00000142 r9:cedaa000 r8:c02011c4 r7:00000142 r6:00000000 r5:b6f5e968
[    8.386685]  r4:004e1f50
[    8.386700] [<c03f6fcc>] (sys_openat) from [<c02011a4>] (__sys_trace_return+0x0/0x1c)
[    8.386709] Exception stack(0xcedabfa8 to 0xcedabff0)
[    8.386721] bfa0:                   004e1f50 b6f5e968 ffffff9c be983f36 00020000 00000000
[    8.386733] bfc0: 004e1f50 b6f5e968 00000000 00000142 00000000 00000000 004e1f50 00000000
[    8.386743] bfe0: b6ef8000 be983c20 004d0ab0 b6e73464
```

naushir commented:
Snap, got exactly the same callstack and could not find an open handle in lsof.  Just adding some extra logging in v4l2 core to see what I can find.

6by9 commented:
Having set vdev->dev_debug = 0xFF, I get logged that the open failed
```
[    8.813304] videodev: v4l2_open: video1: open (-515).
```
Ah, 
```
	ret = v4l2_subdev_call(dev->sensor, core, s_power, 1);
	if (ret < 0 && ret != -ENOIOCTLCMD) {
		unicam_err(dev, "s_power failed\n");
		v4l2_fh_release(file);
		node->open--;
		goto unlock;
	}
```
So in the case where s_power returns ENOIOCTLCMD (515), ret isn't being reset to 0 for success.

6by9 commented:
unicam_open
```
        ret = v4l2_subdev_call(dev->sensor, core, s_power, 1);
-       if (ret < 0 && ret != -ENOIOCTLCMD) {
-               v4l2_fh_release(file);
-               node->open--;
-               goto unlock;
+       if (ret) {
+               if (ret == -ENOIOCTLCMD) {
+                       ret = 0;
+               } else {
+                       unicam_err(dev, "s_power failed\n");
+                       v4l2_fh_release(file);
+                       node->open--;
+                       goto unlock;
+               }
        }
```
Solves this.

naushir commented:
Thanks!!!

Forcing ret = 0 before the return works!

naushir committed:
media: bcm2835-unicam: Use dummy buffer if none have been queued

If no buffer has been queued by a userland application, we use an
internal dummy buffer for the hardware to spin in. This will allow
the driver to release the existing userland buffer back to the
application for processing.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir commented:
Ok, all the above fixed should be pushed + rebased.

naushir commented:
@6by9, if you are happy with this let me know and I'll get @popcornmix to merge.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2367 @@
+
+static void unicam_wr_dma_addr(struct unicam_device *dev, dma_addr_t dmaaddr)
+{
+	unsigned int addr = (unsigned int) dmaaddr;
```
Do we need to cast away from dma_addr_t? The writel within reg_write should do that for us.

6by9 replied:
Checkpatch whinge too
```
CHECK: No space is necessary after a cast
#715: FILE: drivers/media/platform/bcm2835/bcm2835-unicam.c:630:
+	unsigned int addr = (unsigned int) dmaaddr;

total: 0 errors, 1 warnings, 1 checks, 2658 lines checked
```

naushir replied:
The cast was so I could use addr in the log format string with %X.  Otherwise I would have to cast it in the unicam_dbg() call.

If you are against this, I'll change it over.  Otherwise I'll leave it as it is.

Fixed the checkpatch warning though.  Sadly it seems my git precommit hook was not running in all instances, so I missed this warning.

pelwell replied:
You can't get away with treating DMA addresses (or any address) as a u32 any more - it's better to use "%pADx" (the final x forces the address to be printed where otherwise the kernel might hide it for security reasons) and pass in a pointer to the DMA address. This does mean you'll need an intermediate variable for addr + size so you can get a pointer, or you could just print the size.

naushir replied:
Understood, will remove the cast.
EDIT: I'll actually remove the log message - it is not conveying anything useful in normal operation.

6by9 replied:
I was about to add the same thing! The only place the address can be treated as 32 bit is when actually writing to the 32bit register.
And to extend Phil's not, it is passed by reference and hence needs the intermediate variable.

naushir replied:
Done.

6by9 commented on arch/arm/configs/bcm2709_defconfig:
```diff
@@ -874,6 +874,7 @@ CONFIG_VIDEO_EM28XX_V4L2=m
 CONFIG_VIDEO_EM28XX_ALSA=m
 CONFIG_VIDEO_EM28XX_DVB=m
 CONFIG_V4L_PLATFORM_DRIVERS=y
+CONFIG_VIDEO_BCM2835_UNICAM=m
```
Github is howing this defconfig change as part of "media: uapi: v4l2-core: Add sensor ancillary data V4L2 foucc type."

naushir replied:
Oops, that was my fault with the rebasing.

Fixed.

6by9 commented on Documentation/media/uapi/v4l/pixfmt-meta-sensor-data.rst:
```diff
@@ -0,0 +1,32 @@
+<https://mipi.org/specifications/csi-2>`_
+
+The size of the embedded buffer is defined in units of lines with a
+pixel width specified in bytes. This is obtained by a call to the
```
I thought we decided against this as Unicam doesn't parse things based on a width, so it becomes a single line buffer of whatever width makes the buffer big enough.

naushir replied:
Indeed, I will change this around.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -609,44 +626,66 @@ static int unicam_reset_format(struct unicam_node *node)
+			       int pad_id)
 {
 	unsigned int addr = (unsigned int) dmaaddr;
+	unsigned int endaddr = pad_id == IMAGE_PAD ?
```
dma_addr_t again?

naushir replied:
Same as above, If you are against this, I'll change it over.  Otherwise I'll leave it as it is.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1012,6 +1113,106 @@ static int unicam_s_fmt_vid_cap(struct file *file, void *priv,
+	const struct unicam_fmt *fmt = NULL;
+	int ret = 0;
+
+	if (node->pad_id == IMAGE_PAD || f->index != 0)
```
Has something else range checked ```pad_id``` to ensure it isn't greater than the number of pads? If not then why not switch to ```pad_id != METADATA_PAD``` and kill 2 birds with one stone?

naushir replied:
Done.

Should not matter I expect.  The pad_id field is only ever set by us in probe() so the values are explicit unless something has gone horribly wrong.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1012,6 +1113,106 @@ static int unicam_s_fmt_vid_cap(struct file *file, void *priv,
+{
+	struct unicam_node *node = video_drvdata(file);
+
+	if (node->pad_id == IMAGE_PAD)
```
Ditto. ```pad_id != METADATA_PAD```

naushir replied:
Done.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1012,6 +1113,106 @@ static int unicam_s_fmt_vid_cap(struct file *file, void *priv,
+{
+	struct unicam_node *node = video_drvdata(file);
+
+	if (node->pad_id == IMAGE_PAD)
```
Ditto. ```pad_id != METADATA_PAD```

naushir replied:
Done.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1123,17 +1327,31 @@ static void unicam_cfg_image_id(struct unicam_device *dev)
+	/* Do not wrap at the end of the embedded data buffer */
+	set_field(&val, 0, UNICAM_DBOB);
+
+	reg_write(cfg, UNICAM_DCS, val);
```
This looks wrong.  Why is ```val``` set to the register offset at line 1342?
Did you intend to read back the current register value via reg_read first, or something else?

naushir replied:
Oops, that was meant to be a read_reg.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1815,12 +2058,14 @@ static int unicam_open(struct file *file)
 	ret = v4l2_subdev_call(dev->sensor, core, s_power, 1);
 	if (ret < 0 && ret != -ENOIOCTLCMD) {
 		v4l2_fh_release(file);
+		node->open--;
```
Minor nitpick, why not just increment node->open at the end alongside the ret = 0; ?

naushir replied:
Done.

naushir committed:
media: bcm2835-unicam: Add support for mulitple device nodes.

Move device node specific state out of the device state structure and
into a new node structure.  This separation will be needed for future
changes where we will add an embedded data node to the driver to work
alongside the existing image data node.

Currently only use a single image node, so this commit does not add
any functional changes.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

jmondi committed:
Documentation: media: Update sub-device API intro

Update the V4L2 sub-device userspace API introduction to provide more
details on why complex devices might want to register devnodes for the
connected subdevices.

Acked-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Suggested-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Jacopo Mondi <jacopo@jmondi.org>

jmondi committed:
Documentation: media: Document read-only subdevice

Document a new kAPI function to register subdev device nodes in read only
mode and for each affected ioctl report how access is restricted.

Acked-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Jacopo Mondi <jacopo@jmondi.org>

jmondi committed:
media: v4l2-dev: Add v4l2_device_register_ro_subdev_node()

Add to the V4L2 core a function to register device nodes for video
subdevices in read-only mode.

Registering a device node in read-only mode is useful to expose to
userspace the current sub-device configuration, without allowing
application to change it by using the V4L2 subdevice ioctls.

Acked-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Jacopo Mondi <jacopo@jmondi.org>

6by9 commented:
Jacopo has just sent v3 of his patches to the list - https://patchwork.linuxtv.org/project/linux-media/list/?series=2120
v2->v3:
- Add Sakari's ack to the series
- Documentation:
  - Address Sakari' and Hans suggestions
- Implementation:
  - Rename V4L2_FL_RO_DEVNODE to V4L2_FL_SUBDEV_RO_DEVNODE
  - Limit the ability to register sub-device video device nodes to
    driver claiming support for CONFIG_VIDEO_V4L2_SUBDEV_API

Could you update them please? (Sorry!)
I'm in the process of reviewing the rest.

naushir commented:
I was just in the process of getting an update pushed :)

naushir commented:
Jacopo's v3 patchset is now incuded.

6by9 commented:
I'm reviewing it all again. Please don't update until I'm done!

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2369,13 +2420,21 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
+
+	for (i = 0; i < MAX_NODES; i++) {
+		node = &unicam->node[i];
+		if (node->registered) {
```
If video_register_device failed in register_node, then we have done the dma_alloc_coherent but not set node->registered. The cleanup will therefore omit freeing the allocation.

6by9 replied:
And a more general comment before upstreaming, we ought to support unbind too so that a subdev can be unloaded. The ctrl_handler code seems to be a little tangled up between probe and bound in the current driver, and I fear potentially leaves old controls lying around.

naushir replied:
Done the video_register_device failure case.

Done the list ctrl on node 1 fix.  The ctrl_handler was set globally for the driver and not per node.  I had actually changed that some time back, but that change seems to have gotten lost in the process.

Regarding the bind/unbind, not done anything for this yet, but I wonder if the ctrl_handler registration should be moved into node open() instead of probe?

6by9 replied:
> Regarding the bind/unbind, not done anything for this yet, but I wonder if the ctrl_handler registration should be moved into node open() instead of probe?

Probably, but I'm not going to worry for now. I've had enough of reading through this driver! I'd need to investigate how unbind is triggered anyway. IIRC Unicam will take a reference on the subdev (or the other way around), so you have to unload them in the correct order, which I think is Unicam first.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -652,27 +693,55 @@ static inline unsigned int unicam_get_lines_done(struct unicam_device *dev)
+
+	ret = dev->node[IMAGE_PAD].open && !dev->node[IMAGE_PAD].streaming;
+	ret &= !dev->node[METADATA_PAD].open ||
+	       !dev->node[METADATA_PAD].streaming;
```
Can't this be simplified to ```!dev->node[IMAGE_PAD].streaming && !dev->node[METADATA_PAD].streaming```?
If the metadata node isn't opened, then it can't be streaming, so we're all disabled anyway. (Might be worth double checking that release does clear streaming under all circumstances).

naushir replied:
Done.

To be safe I did add a clause in release() to stop streaming if the device was asked to close and we were still streaming.

6by9 reviewed (commented):
A couple of minor comments, and a weird observation.

Running with an unmodified imx219, ```v4l2-ctl --list-ctrls -d 1``` lists all the controls for imx219, exactly as if it had been the video node. Is something sharing a ctrl_handle pointer when it shouldn't be?

naushir commented:
@6by9, did you get a chance to go through the last set of fixes?  I'm readying the PRs for imx219 and ISP that need to sit on top of these changes.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1814,12 +2056,22 @@ static int unicam_open(struct file *file)
+	 * struct v4l2_fh gets allocated in v4l2_fh_open() above.
+	 */
+	if (node->pad_id == IMAGE_PAD) {
+		fh = (struct v4l2_fh *)file->private_data;
```
Having to cast a private_data pointer seems like the wrong thing to me.
I had a thread with Hans about where to set ctrl_handlers  - https://www.spinics.net/lists/arm-kernel/msg779301.html

quote: You can set ctrl_handlers at three levels: v4l2_device (common to all
video devices), video_device (specific to a single video device), or
for v4l2_fh (specific to that filehandle only).

In this case you would assign it to the video_device:

	dev->vdev.ctrl_handler = &dev->ctrl_handler;
(end quote)

I think you need the same thing here. register_node should be setting ctrl_handler on ```node->video_dev```, not leaving it to be set on the file handle.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -687,18 +752,20 @@ static irqreturn_t unicam_isr(int irq, void *dev)
 	unsigned int lines_done = unicam_get_lines_done(dev);
 	unsigned int sequence = unicam->sequence;
+	int num_nodes_streaming = unicam_num_nodes_streaming(dev);
+	u64 ts;
```
I wasn't going to nit-pick over this one, but have had to raise issue below. Longest variable description to shortest, so ts should be after ista,sta.

naushir replied:
Done.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1996,13 +2268,20 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
+		 * ctrl handler.
+		 */
+		if (list_empty(&node->ctrl_handler.ctrls))
+			unicam->v4l2_dev.ctrl_handler = NULL;
```
Is this the right thing to be NULLing? I didn't think it was set, and you've been setting it fh_open.

6by9 replied:
Actually I think you want
```
		if (!list_empty(&node->ctrl_handler.ctrls))
			vdev->ctrl_handler = &node->ctrl_handler;
```
having first moved the assignment of vdev from a dozen lines later to just before this.

naushir replied:
Done.

6by9 reviewed (changes requested):
A diff of 
```
--- a/drivers/media/platform/bcm2835/bcm2835-unicam.c
+++ b/drivers/media/platform/bcm2835/bcm2835-unicam.c
@@ -2084,7 +2084,6 @@ static int unicam_open(struct file *file)
 {
        struct unicam_node *node = video_drvdata(file);
        struct unicam_device *dev = node->dev;
-       struct v4l2_fh *fh;
        int ret;
 
        mutex_lock(&node->lock);
@@ -2098,15 +2097,6 @@ static int unicam_open(struct file *file)
        if (!v4l2_fh_is_singular_file(file))
                goto unlock;
 
-       /*
-        * Ensure only the image node assigns a ctrl handler.
-        * struct v4l2_fh gets allocated in v4l2_fh_open() above.
-        */
-       if (node->pad_id == IMAGE_PAD) {
-               fh = (struct v4l2_fh *)file->private_data;
-               fh->ctrl_handler = &node->ctrl_handler;
-       }
-
        ret = v4l2_subdev_call(dev->sensor, core, s_power, 1);
        if (ret < 0 && ret != -ENOIOCTLCMD) {
                v4l2_fh_release(file);
@@ -2310,6 +2300,7 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
        spin_lock_init(&node->dma_queue_lock);
        mutex_init(&node->lock);
 
+       vdev = &node->video_dev;
        if (pad_id == IMAGE_PAD) {
                /* Add controls from the subdevice */
                ret = v4l2_ctrl_add_handler(&node->ctrl_handler,
@@ -2322,8 +2313,8 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
                 * If the source has no controls then remove our
                 * ctrl handler.
                 */
-               if (list_empty(&node->ctrl_handler.ctrls))
-                       unicam->v4l2_dev.ctrl_handler = NULL;
+               if (!list_empty(&node->ctrl_handler.ctrls))
+                       vdev->ctrl_handler = &node->ctrl_handler;
        }
 
        q = &node->buffer_queue;
@@ -2346,7 +2337,6 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
 
        INIT_LIST_HEAD(&node->dma_queue.active);
 
-       vdev = &node->video_dev;
        vdev->release = video_device_release_empty;
        vdev->fops = &unicam_fops;
        vdev->ioctl_ops = &unicam_ioctl_ops;
```
seems to do the right thing for me.

naushir commented:
Thanks, I've Incorporated your diff for the ctrl handler and all looks good to me.

6by9 commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2782 @@
+			return ret;
+
+		/*
+		 * If the source has no controls then remove our
```
Sorry, I didn't amend the comment in my diff.

naushir replied:
Missed that as well.  Now fixed.

6by9 commented:
Assuming that the comment was the only thing changed in the last patch set, then I'm happy. There are still a couple of odd bits, but we need to draw a line somewhere.

naushir commented:
> Assuming that the comment was the only thing changed in the last patch set, then I'm happy. There are still a couple of odd bits, but we need to draw a line somewhere.

Yes, we still have to rework the probe/bound functions and using g_mbus_format pad ops before upstreaming.  However, pushing this unblocks the other 2 PRs that I have pending :)

@pelwell or @popcornmix, could you merge if you are happy please?

pelwell commented on MAINTAINERS:
```diff
@@ -3206,7 +3206,7 @@ F:	Documentation/devicetree/bindings/media/rpivid_hevc.jaml
 F:	drivers/staging/media/rpivid
```
There's a repeated "of" in the commit message, but I can live with it if you can.

naushir replied:
Done

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2363 @@
+
+static void unicam_wr_dma_addr(struct unicam_device *dev, dma_addr_t dmaaddr)
+{
+	reg_write(&dev->cfg, UNICAM_IBSA0, dmaaddr);
```
I think we should acknowledge the possibility that dmaaddr is a 64-bit quantity but that these are 32-bit registers (or is it effectively 30-bit?), either with a comment or a BUG/WARN if an out-of-range address is used.

naushir replied:
It's 32-bits, but the top 2-bits are set to 0x3 for uncached access.  I've put in a BUG_ON(...) for this test.

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -708,14 +721,29 @@ static inline void unicam_schedule_next_buffer(struct unicam_node *node)
+	dma_addr_t addr = node->dummy_buf_dma_addr;
+
+	if (node->pad_id == IMAGE_PAD) {
+		reg_write(&dev->cfg, UNICAM_IBSA0, addr);
```
This is so close to `unicam_wr_dma_addr`, can't you make use of it?

naushir replied:
Done.

naushir committed:
media: bcm2835-unicam: Add embedded data node.

This patch adds a new node in the bcm2835-unicam driver to support
CSI-2 embedded data streams.  The subdevice is queried to see if
embedded data is available from the sensor.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>
